File: app/src/main/java/com/origin/launcher/HomeFragment.java
package com.origin.launcher;

import android.content.DialogInterface;
import org.jetbrains.annotations.NotNull;
import android.annotation.SuppressLint;
import android.app.AlertDialog;
import android.os.Handler;
import android.os.Looper;
import android.view.View;
import android.widget.TextView;
import android.widget.Button;
import android.widget.EditText;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import androidx.core.content.FileProvider;

import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.os.Build;
import androidx.fragment.app.Fragment;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.FileOutputStream;
import java.io.FileInputStream;
import java.io.FileWriter;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Objects;
import java.util.concurrent.Executors;
import java.util.zip.ZipEntry;
import java.util.zip.ZipFile;
import java.util.zip.ZipInputStream;
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.util.ArrayList;
import android.os.Looper;
import android.content.res.ColorStateList;
import android.graphics.Color;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.card.MaterialCardView;
import com.origin.launcher.ThemeManager;
import com.origin.launcher.ThemeUtils;
import android.util.Log;
import android.widget.Toast;
import androidx.dynamicanimation.animation.SpringAnimation;
import androidx.dynamicanimation.animation.SpringForce;
import com.origin.launcher.Launcher.MinecraftLauncher;
import com.origin.launcher.versions.GameVersion;
import com.origin.launcher.versions.VersionManager;
import com.origin.launcher.databinding.ActivityMainBinding;
import com.origin.launcher.FeatureSettings;
import com.origin.launcher.animation.DynamicAnim;

public class HomeFragment extends BaseThemedFragment {

    private static final String TAG = "HomeFragment";
    private TextView listener;
    private Button mbl2_button;
    private Button versions_button;
    private com.google.android.material.button.MaterialButton shareLogsButton;
    private MinecraftLauncher minecraftLauncher;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_home, container, false);

        listener = view.findViewById(R.id.listener);
        mbl2_button = view.findViewById(R.id.mbl2_load);
        versions_button = view.findViewById(R.id.versions_button);
        shareLogsButton = view.findViewById(R.id.share_logs_button);
        Handler handler = new Handler(Looper.getMainLooper());

        // Apply initial theme
        applyInitialTheme(view);
        
        mbl2_button.setOnClickListener(v -> launchGame());

        // Long press to clear APK selection
        mbl2_button.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                clearSelectedApk();
                return true;
            }
        });

        versions_button.setOnClickListener(v -> {
            try {
                requireActivity().getSupportFragmentManager()
                    .beginTransaction()
                    .setCustomAnimations(
                        R.anim.slide_fade_in_right,  
                        R.anim.slide_out_right, 
                        R.anim.slide_in_left,   
                        R.anim.slide_out_left 
                    )
                    .replace(android.R.id.content, new VersionsFragment())
                    .addToBackStack(null)
                    .commit();

                Log.d(TAG, "Opening themes fragment");
            } catch (Exception e) {
                Log.e(TAG, "Error opening themes", e);
                Toast.makeText(getContext(), "Unable to open themes", Toast.LENGTH_SHORT).show();
            }
        });

        // Set initial log text
        listener.setText("Ready to launch Minecraft");

        // Show current selection status
        updateSelectionStatus();

        // Set up share button
        shareLogsButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                shareLogs();
            }
        });

        return view;
    }

    /**
     * Apply initial theme to all views
     */
    private void applyInitialTheme(View view) {
        try {
            ThemeManager themeManager = ThemeManager.getInstance();
            if (themeManager != null && themeManager.isThemeLoaded()) {
                // Apply theme to main button
                if (mbl2_button instanceof MaterialButton) {
                    ThemeUtils.applyThemeToButton((MaterialButton) mbl2_button, requireContext());
                }

                // Apply theme to share button (remove background, make it text button)
                if (shareLogsButton != null) {
                    ThemeUtils.applyThemeToButton(shareLogsButton, requireContext());
                    // Remove background and make it transparent
                    shareLogsButton.setBackgroundTintList(ColorStateList.valueOf(Color.TRANSPARENT));
                    shareLogsButton.setStrokeWidth(0);
                }

                // Ensure versions nav button is transparent like share button
                if (versions_button instanceof MaterialButton) {
                    MaterialButton vb = (MaterialButton) versions_button;
                    ThemeUtils.applyThemeToButton(vb, requireContext());
                    vb.setBackgroundTintList(ColorStateList.valueOf(Color.TRANSPARENT));
                    vb.setStrokeWidth(0);
                    try {
                        vb.setIconTint(ColorStateList.valueOf(themeManager.getColor("onSurfaceVariant")));
                    } catch (Exception ignored) {}
                }

                // Apply theme to log text area
                if (listener != null) {
                    listener.setTextColor(themeManager.getColor("onSurfaceVariant"));
                    // Set background color for the log text area
                    View logCard = view.findViewById(R.id.logCard);
                    if (logCard instanceof MaterialCardView) {
                        MaterialCardView card = (MaterialCardView) logCard;
                        card.setCardBackgroundColor(themeManager.getColor("surfaceVariant"));
                        card.setStrokeColor(themeManager.getColor("outline"));
                    }
                }
            }
        } catch (Exception e) {
            // Handle error gracefully
        }
    }
    
  private void setupManagersAndHandlers() {
        versionManager.loadAllVersions();
        minecraftLauncher = new MinecraftLauncher(this);
    }

   private void checkResourcepack() {
        ExecutorService executorService = Executors.newSingleThreadExecutor();
        new ResourcepackHandler(
                this, minecraftLauncher, executorService
        ).checkIntentForResourcepack();
    }

    private void launchGame() {
        binding.mbl2_button.setEnabled(false);

        GameVersion version = versionManager != null ? versionManager.getSelectedVersion() : null;

        if (version == null) {
            binding.mbl2_button.setEnabled(true);
            new CustomAlertDialog(this)
                    .setTitleText(getString(R.string.dialog_title_no_version))
                    .setMessage(getString(R.string.dialog_message_no_version))
                    .setPositiveButton(getString(R.string.dialog_positive_ok), null)
                    .show();
            return;
        }

        if (!version.isInstalled && !FeatureSettings.getInstance().isVersionIsolationEnabled()) {
            binding.mbl2_button.setEnabled(true);
            new CustomAlertDialog(this)
                    .setTitleText(getString(R.string.dialog_title_version_isolation))
                    .setMessage(getString(R.string.dialog_message_version_isolation))
                    .setPositiveButton(getString(R.string.dialog_positive_enable), v -> {
                        FeatureSettings.getInstance().setVersionIsolationEnabled(true);
                        launchGame();
                    })
                    .setNegativeButton(getString(R.string.dialog_negative_cancel), null)
                    .show();
            return;
        }
        new Thread(() -> {
            try {
                minecraftLauncher.launch(getIntent(), version);
                runOnUiThread(() -> {
                    binding.mbl2_button.setEnabled(true);
                });
            } catch (Exception e) {
                runOnUiThread(() -> {
                    binding.mbl2_button.setEnabled(true);
                    new CustomAlertDialog(this)
                            .setTitleText(getString(R.string.dialog_title_launch_failed))
                            .setMessage(getString(R.string.dialog_message_launch_failed, e.getMessage()))
                            .setPositiveButton(getString(R.string.dialog_positive_ok), null)
                            .show();
                });
            }
        }).start();
    }
  
      private boolean animatembl2_button(View v, MotionEvent event) {
        if (event.getAction() == MotionEvent.ACTION_DOWN) {
            SpringAnimation sx = new SpringAnimation(v, SpringAnimation.SCALE_X, 0.95f);
            SpringAnimation sy = new SpringAnimation(v, SpringAnimation.SCALE_Y, 0.95f);
            SpringForce spring = new SpringForce(0.95f)
                    .setDampingRatio(SpringForce.DAMPING_RATIO_NO_BOUNCY)
                    .setStiffness(SpringForce.STIFFNESS_MEDIUM);
            sx.setSpring(spring);
            sy.setSpring(spring);
            sx.start();
            sy.start();
        } else if (event.getAction() == MotionEvent.ACTION_UP || event.getAction() == MotionEvent.ACTION_CANCEL) {
            SpringAnimation sx = new SpringAnimation(v, SpringAnimation.SCALE_X, 1f);
            SpringAnimation sy = new SpringAnimation(v, SpringAnimation.SCALE_Y, 1f);
            SpringForce spring = new SpringForce(1f)
                    .setDampingRatio(SpringForce.DAMPING_RATIO_MEDIUM_BOUNCY)
                    .setStiffness(SpringForce.STIFFNESS_LOW);
            sx.setSpring(spring);
            sy.setSpring(spring);
            sx.start();
            sy.start();
        }
        return false;
    }

    @Override
    protected void onApplyTheme() {
        super.onApplyTheme();

        View view = getView();
        if (view != null) {
            // Refresh all theme elements
            applyInitialTheme(view);
        }
    }

    private String getPackageNameFromSettings() {
        SharedPreferences prefs = requireContext().getSharedPreferences("settings", 0);
        return prefs.getString("mc_package_name", "com.mojang.minecraftpe");
    }

    private String getSelectedApkPath() {
        SharedPreferences prefs = requireContext().getSharedPreferences("selected_apk", 0);
        return prefs.getString("apk_path", null);
    }

    private void updateSelectionStatus() {
        String selectedApkPath = getSelectedApkPath();
        if (selectedApkPath != null && new File(selectedApkPath).exists()) {
            String fileName = new File(selectedApkPath).getName();
            listener.setText("Ready to launch Minecraft\nSelected APK: " + fileName);
        } else {
            listener.setText("Ready to launch Minecraft");
        }
    }

    private void clearSelectedApk() {
        SharedPreferences prefs = requireContext().getSharedPreferences("selected_apk", 0);
        prefs.edit().remove("apk_path").apply();
        updateSelectionStatus();
        Toast.makeText(requireContext(), "Cleared APK selection", Toast.LENGTH_SHORT).show();
    }

    private void shareLogs() {
        try {
            // Get the current log text
            String logText = listener.getText().toString();

            // Create a temporary file
            File logFile = new File(requireContext().getCacheDir(), "latestlog.txt");
            FileWriter writer = new FileWriter(logFile);
            writer.write(logText);
            writer.close();

            // Create the sharing intent
            Intent shareIntent = new Intent(Intent.ACTION_SEND);
            shareIntent.setType("text/plain");

            // Get the file URI using FileProvider
            android.net.Uri fileUri = FileProvider.getUriForFile(
                requireContext(),
                "com.origin.launcher.fileprovider",
                logFile
            );

            shareIntent.putExtra(Intent.EXTRA_STREAM, fileUri);
            shareIntent.putExtra(Intent.EXTRA_SUBJECT, "Xelo Client Logs");
            shareIntent.putExtra(Intent.EXTRA_TEXT, "Xelo Client Latest Logs");
            shareIntent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);

            // Start the sharing activity
            startActivity(Intent.createChooser(shareIntent, "Share Logs"));

        } catch (Exception e) {
            // Show error message
            android.widget.Toast.makeText(requireContext(), "Failed to share logs: " + e.getMessage(), android.widget.Toast.LENGTH_SHORT).show();
        }
    }

    @SuppressLint("SetTextI18n")

    private Object getPathList(@NotNull ClassLoader classLoader) throws Exception {
        Field pathListField = Objects.requireNonNull(classLoader.getClass().getSuperclass()).getDeclaredField("pathList");
        pathListField.setAccessible(true);
        return pathListField.get(classLoader);
    }

    private boolean processNativeLibraries(ApplicationInfo mcInfo, @NotNull Object pathList, @NotNull Handler handler, TextView listener) throws Exception {
        FileInputStream inStream = new FileInputStream(getApkWithLibs(mcInfo));
        BufferedInputStream bufInStream = new BufferedInputStream(inStream);
        ZipInputStream inZipStream = new ZipInputStream(bufInStream);
        if (!checkLibCompatibility(inZipStream)) {
            handler.post(() -> alertAndExit("Wrong minecraft architecture", "The minecraft you have installed does not support the same main architecture (" + Build.SUPPORTED_ABIS[0] + ") your device uses, Xelo client cant work with it"));
            return false;
        }                     
        Method addNativePath = pathList.getClass().getDeclaredMethod("addNativePath", Collection.class);
        ArrayList<String> libDirList = new ArrayList<>();
        File libdir = new File(mcInfo.nativeLibraryDir);
        if (libdir.list() == null || libdir.list().length == 0 
         || (mcInfo.flags & ApplicationInfo.FLAG_EXTRACT_NATIVE_LIBS) != ApplicationInfo.FLAG_EXTRACT_NATIVE_LIBS) {
            loadUnextractedLibs(mcInfo);
            libDirList.add(requireActivity().getCodeCacheDir().getAbsolutePath() + "/");
        } else {
            libDirList.add(mcInfo.nativeLibraryDir);
        }
        addNativePath.invoke(pathList, libDirList);
        handler.post(() -> listener.append("\n-> " + mcInfo.nativeLibraryDir + " added to native library directory path"));
        return true;
    }

    private static Boolean checkLibCompatibility(ZipInputStream zip) throws Exception{
         ZipEntry ze = null;
         String requiredLibDir = "lib/" + Build.SUPPORTED_ABIS[0] + "/";
         while ((ze = zip.getNextEntry()) != null) {
             if (ze.getName().startsWith(requiredLibDir)) {
                 return true;
             }
         }
         zip.close();
         return false;
     }

     private void alertAndExit(String issue, String description) {
        AlertDialog alertDialog = new AlertDialog.Builder(requireActivity()).create();
        alertDialog.setTitle(issue);
        alertDialog.setMessage(description);
        alertDialog.setCancelable(false);
        alertDialog.setButton(AlertDialog.BUTTON_NEUTRAL, "Exit",
        new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int which) {
                requireActivity().finish();
            }
        });
        alertDialog.show();         
     }

    private void loadUnextractedLibs(ApplicationInfo appInfo) throws Exception {
        FileInputStream inStream = new FileInputStream(getApkWithLibs(appInfo));
        BufferedInputStream bufInStream = new BufferedInputStream(inStream);
        ZipInputStream inZipStream = new ZipInputStream(bufInStream);
        String zipPath = "lib/" + Build.SUPPORTED_ABIS[0] + "/";
        String outPath = requireActivity().getCodeCacheDir().getAbsolutePath() + "/";
        File dir = new File(outPath);
        dir.mkdir();
        extractDir(appInfo, inZipStream, zipPath, outPath);
    }

    public String getApkWithLibs(ApplicationInfo pkg) throws PackageManager.NameNotFoundException {
        String[] sn=pkg.splitSourceDirs;
        if (sn != null && sn.length > 0) {
            String cur_abi = Build.SUPPORTED_ABIS[0].replace('-','_');
            for(String n:sn){
                if(n.contains(cur_abi)){
                    return n;
                }
            }
        }
        return pkg.sourceDir;
    }

    private static void extractDir(ApplicationInfo mcInfo, ZipInputStream zip, String zip_folder, String out_folder ) throws Exception{
        ZipEntry ze = null;
        while ((ze = zip.getNextEntry()) != null) {
            if (ze.getName().startsWith(zip_folder) && !ze.getName().contains("c++_shared")) {
                String strippedName = ze.getName().substring(zip_folder.length());
                String path = out_folder + "/" + strippedName;
                OutputStream out = new FileOutputStream(path);
                BufferedOutputStream outBuf = new BufferedOutputStream(out);
                byte[] buffer = new byte[9000];
                int len;
                while ((len = zip.read(buffer)) != -1) {
                    outBuf.write(buffer, 0, len);
                }
                outBuf.close();
            }
        }
        zip.close();
    }

    private static void copyFile(InputStream from, @NotNull File to) throws IOException {
        File parentDir = to.getParentFile();
        if (parentDir != null && !parentDir.exists() && !parentDir.mkdirs()) {
            throw new IOException("Failed to create directories");
        }
        if (!to.exists() && !to.createNewFile()) {
            throw new IOException("Failed to create new file");
        }
        try (BufferedInputStream input = new BufferedInputStream(from);
             BufferedOutputStream output = new BufferedOutputStream(Files.newOutputStream(to.toPath()))) {
            byte[] buffer = new byte[8192];
            int bytesRead;
            while ((bytesRead = input.read(buffer)) != -1) {
                output.write(buffer, 0, bytesRead);
            }
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        DiscordRPCHelper.getInstance().updateMenuPresence("Playing");
    }

    @Override
    public void onPause() {
        super.onPause();
        DiscordRPCHelper.getInstance().updateIdlePresence();
    }
}

File: app/src/main/java/com/origin/launcher/HomeFragment.kt
package com.origin.launcher

import android.annotation.SuppressLint
import android.app.AlertDialog
import android.content.DialogInterface
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.content.res.ColorStateList
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.core.content.FileProvider
import androidx.fragment.app.Fragment
import com.google.android.material.button.MaterialButton
import com.google.android.material.card.MaterialCardView
import com.origin.launcher.ThemeManager
import com.origin.launcher.ThemeUtils
import com.origin.launcher.LoadingDialog
import java.io.*
import java.nio.file.Files
import java.util.*
import java.util.concurrent.Executors
import java.util.zip.ZipEntry
import java.util.zip.ZipFile
import java.util.zip.ZipInputStream
import kotlin.collections.ArrayList
import android.view.animation.AnimationUtils
import android.widget.ImageView
import com.origin.launcher.Launcher.MinecraftLauncher

class HomeFragment : BaseThemedFragment() {

    companion object {
        private const val TAG = "HomeFragment"
    }

    private lateinit var listener: TextView
    private lateinit var mbl2_button: Button
    private lateinit var versions_button: Button
    private lateinit var shareLogsButton: MaterialButton
private var loadingDialog: LoadingDialog? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val view = inflater.inflate(R.layout.fragment_home, container, false)

        listener = view.findViewById(R.id.listener)
        mbl2_button = view.findViewById(R.id.mbl2_load)
        versions_button = view.findViewById(R.id.versions_button)
        shareLogsButton = view.findViewById(R.id.share_logs_button)
        val handler = Handler(Looper.getMainLooper())

        applyInitialTheme(view)

        mbl2_button.setOnClickListener {
    mbl2_button.isEnabled = false
    listener.text = "Starting Minecraft launcher..."
    
    val handler = Handler(Looper.getMainLooper())
    val packageName = getPackageNameFromSettings()
    
    Executors.newSingleThreadExecutor().execute {
        try {
            val mcInfo = getMinecraftInfo(packageName)
            val pathList = getPathList(requireActivity().classLoader)
            
            if (processNativeLibraries(mcInfo, pathList, handler, listener)) {
                handler.postDelayed({
                    try {
                        val intent = Intent(requireContext(), MinecraftLauncher::class.java)
                        intent.putExtra("MODS_ENABLED", true)
                        startActivity(intent)
listener.append("""
-> MinecraftActivity launching!
""".trimIndent())
                    } catch (e: Exception) {
                        listener.text = "Activity launch failed: ${e.message}"
                        mbl2_button.isEnabled = true
                    }
                }, 1000)
            } else {
                handler.post { mbl2_button.isEnabled = true }
            }
        } catch (e: Exception) {
            handler.post {
                listener.text = "Launch failed: ${e.message}"
                mbl2_button.isEnabled = true
            }
        }
    }
}


        mbl2_button.setOnLongClickListener {
            clearSelectedApk()
            true
        }

        versions_button.setOnClickListener {
            try {
                requireActivity().supportFragmentManager
                    .beginTransaction()
                    .setCustomAnimations(
                        R.anim.slide_fade_in_right,
                        R.anim.slide_out_right,
                        R.anim.slide_in_left,
                        R.anim.slide_out_left
                    )
                    .replace(android.R.id.content, VersionsFragment())
                    .addToBackStack(null)
                    .commit()

                Log.d(TAG, "Opening themes fragment")
            } catch (e: Exception) {
                Log.e(TAG, "Error opening themes", e)
                Toast.makeText(context, "Unable to open themes", Toast.LENGTH_SHORT).show()
            }
        }

        listener.text = "Ready to launch Minecraft"
        updateSelectionStatus()

        shareLogsButton.setOnClickListener { shareLogs() }

        return view
    }

private fun getMinecraftInfo(packageName: String): ApplicationInfo {
    val selected = getSelectedApkPath()
    return if (selected != null && File(selected).exists()) {
        val mcInfo = requireActivity().packageManager.getApplicationInfo(packageName, 0)
        mcInfo.sourceDir = selected
        mcInfo
    } else {
        requireActivity().packageManager.getApplicationInfo(packageName, 0)
    }
}

    private fun applyInitialTheme(view: View) {
        try {
            val themeManager = ThemeManager.getInstance()
            if (themeManager != null && themeManager.isThemeLoaded) {

                if (mbl2_button is MaterialButton) {
                    ThemeUtils.applyThemeToButton(mbl2_button as MaterialButton, requireContext())
                }

                ThemeUtils.applyThemeToButton(shareLogsButton, requireContext())
                shareLogsButton.backgroundTintList = ColorStateList.valueOf(Color.TRANSPARENT)
                shareLogsButton.strokeWidth = 0

                if (versions_button is MaterialButton) {
                    val vb = versions_button as MaterialButton
                    ThemeUtils.applyThemeToButton(vb, requireContext())
                    vb.backgroundTintList = ColorStateList.valueOf(Color.TRANSPARENT)
                    vb.strokeWidth = 0
                    try {
                        vb.iconTint = ColorStateList.valueOf(themeManager.getColor("onSurfaceVariant"))
                    } catch (_: Exception) {}
                }

                listener.setTextColor(themeManager.getColor("onSurfaceVariant"))

                val logCard = view.findViewById<View>(R.id.logCard)
                if (logCard is MaterialCardView) {
                    logCard.setCardBackgroundColor(themeManager.getColor("surfaceVariant"))
                    logCard.strokeColor = themeManager.getColor("outline")
                }
            }
        } catch (_: Exception) {
        }
    }

    override fun onApplyTheme() {
        super.onApplyTheme()
        view?.let { applyInitialTheme(it) }
    }

    private fun getPackageNameFromSettings(): String {
        val prefs = requireContext().getSharedPreferences("settings", 0)
        return prefs.getString("mc_package_name", "com.mojang.minecraftpe")!!
    }

    private fun getSelectedApkPath(): String? {
        val prefs = requireContext().getSharedPreferences("selected_apk", 0)
        return prefs.getString("apk_path", null)
    }

    private fun updateSelectionStatus() {
        val path = getSelectedApkPath()
        if (path != null && File(path).exists()) {
            val name = File(path).name
            listener.text = "Ready to launch Minecraft\nSelected APK: $name"
        } else {
            listener.text = "Ready to launch Minecraft"
        }
    }

    private fun clearSelectedApk() {
        val prefs = requireContext().getSharedPreferences("selected_apk", 0)
        prefs.edit().remove("apk_path").apply()
        updateSelectionStatus()
        Toast.makeText(requireContext(), "Cleared APK selection", Toast.LENGTH_SHORT).show()
    }

    private fun shareLogs() {
        try {
            val logText = listener.text.toString()
            val logFile = File(requireContext().cacheDir, "latestlog.txt")

            FileWriter(logFile).apply {
                write(logText)
                close()
            }

            val fileUri = FileProvider.getUriForFile(
                requireContext(),
                "com.origin.launcher.fileprovider",
                logFile
            )

            val intent = Intent(Intent.ACTION_SEND).apply {
                type = "text/plain"
                putExtra(Intent.EXTRA_STREAM, fileUri)
                putExtra(Intent.EXTRA_SUBJECT, "Xelo Client Logs")
                putExtra(Intent.EXTRA_TEXT, "Xelo Client Latest Logs")
                addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
            }

            startActivity(Intent.createChooser(intent, "Share Logs"))

        } catch (e: Exception) {
            Toast.makeText(requireContext(), "Failed to share logs: ${e.message}", Toast.LENGTH_SHORT).show()
        }
    }

    

    private fun getPathList(classLoader: ClassLoader): Any {
        val field = classLoader.javaClass.superclass!!.getDeclaredField("pathList")
        field.isAccessible = true
        return field.get(classLoader)
    }

    private fun processNativeLibraries(
        mcInfo: ApplicationInfo,
        pathList: Any,
        handler: Handler,
        listener: TextView
    ): Boolean {
        val zip = ZipInputStream(FileInputStream(getApkWithLibs(mcInfo)))

        if (!checkLibCompatibility(zip)) {
            handler.post {
                alertAndExit(
                    "Wrong architecture",
                    "Minecraft ABI doesn't match device (${Build.SUPPORTED_ABIS[0]})"
                )
            }
            return false
        }

        val addNativePath = pathList.javaClass.getDeclaredMethod("addNativePath", Collection::class.java)

        val libDirs = ArrayList<String>()

        val libdir = File(mcInfo.nativeLibraryDir)

        if (libdir.list() == null ||
            libdir.list().isEmpty() ||
            (mcInfo.flags and ApplicationInfo.FLAG_EXTRACT_NATIVE_LIBS) != ApplicationInfo.FLAG_EXTRACT_NATIVE_LIBS
        ) {
            loadUnextractedLibs(mcInfo)
            libDirs.add(requireActivity().codeCacheDir.absolutePath + "/")
        } else {
            libDirs.add(mcInfo.nativeLibraryDir)
        }

        addNativePath.invoke(pathList, libDirs)

        handler.post { listener.append("\n-> ${mcInfo.nativeLibraryDir} added to native dirs") }
        return true
    }

    private fun checkLibCompatibility(zip: ZipInputStream): Boolean {
        val req = "lib/${Build.SUPPORTED_ABIS[0]}/"
        var entry: ZipEntry?

        while (true) {
            entry = zip.nextEntry ?: break
            if (entry.name.startsWith(req)) return true
        }

        zip.close()
        return false
    }

    private fun alertAndExit(title: String, msg: String) {
        val dialog = AlertDialog.Builder(requireActivity()).create()
        dialog.setTitle(title)
        dialog.setMessage(msg)
        dialog.setCancelable(false)
        dialog.setButton(AlertDialog.BUTTON_NEUTRAL, "Exit") { _: DialogInterface?, _: Int ->
            requireActivity().finish()
        }
        dialog.show()
    }

    private fun loadUnextractedLibs(appInfo: ApplicationInfo) {
        val zip = ZipInputStream(FileInputStream(getApkWithLibs(appInfo)))
        val zipPath = "lib/${Build.SUPPORTED_ABIS[0]}/"
        val out = requireActivity().codeCacheDir.absolutePath + "/"

        File(out).mkdir()
        extractDir(appInfo, zip, zipPath, out)
    }

    fun getApkWithLibs(pkg: ApplicationInfo): String {
        pkg.splitSourceDirs?.let {
            val abi = Build.SUPPORTED_ABIS[0].replace('-', '_')
            for (path in it) {
                if (path.contains(abi)) return path
            }
        }
        return pkg.sourceDir
    }

    private fun extractDir(
        mcInfo: ApplicationInfo,
        zip: ZipInputStream,
        zip_folder: String,
        out_folder: String
    ) {
        var entry: ZipEntry?

        while (true) {
            entry = zip.nextEntry ?: break
            if (entry.name.startsWith(zip_folder) && !entry.name.contains("c++_shared")) {
                val stripped = entry.name.substring(zip_folder.length)
                val out = File(out_folder + stripped)
                val outStream = BufferedOutputStream(FileOutputStream(out))

                val buffer = ByteArray(9000)
                var len: Int

                while (zip.read(buffer).also { len = it } != -1) {
                    outStream.write(buffer, 0, len)
                }

                outStream.close()
            }
        }

        zip.close()
    }

    override fun onResume() {
        super.onResume()
loadingDialog?.dismiss()
    loadingDialog = null
        DiscordRPCHelper.getInstance().updateMenuPresence("Playing")
    }

    override fun onPause() {
        super.onPause()
        DiscordRPCHelper.getInstance().updateIdlePresence()
    }
}